File: app/Services/GraphService.php
<?php

namespace App\Services;

class GraphService
{
    private array $graph = [];
    private array $cities = [];
    private array $distanceMatrix = [];
    private array $pathMatrix = [];

    public function __construct(array $edges)
    {
        foreach ($edges as $edge) {
            [$u, $v, $w] = $edge;
            $this->graph[$u][] = ['city' => $v, 'weight' => $w];
            $this->graph[$v][] = ['city' => $u, 'weight' => $w];
            if (!in_array($u, $this->cities)) $this->cities[] = $u;
            if (!in_array($v, $this->cities)) $this->cities[] = $v;
        }
        $this->computeAllShortestPaths();
    }

    /** Dijkstra dari src ke semua kota lain */
    private function dijkstra(string $src): array
    {
        $dist = [];
        $prev = [];
        $visited = [];
        foreach ($this->cities as $c) {
            $dist[$c] = INF;
            $prev[$c] = null;
            $visited[$c] = false;
        }
        $dist[$src] = 0;

        for ($i = 0; $i < count($this->cities); $i++) {
            $min = INF;
            $u = null;
            foreach ($this->cities as $c) {
                if (!$visited[$c] && $dist[$c] < $min) {
                    $min = $dist[$c];
                    $u = $c;
                }
            }
            if ($u === null) break;
            $visited[$u] = true;
            foreach ($this->graph[$u] ?? [] as $neighbor) {
                $alt = $dist[$u] + $neighbor['weight'];
                if ($alt < $dist[$neighbor['city']]) {
                    $dist[$neighbor['city']] = $alt;
                    $prev[$neighbor['city']] = $u;
                }
            }
        }
        return ['dist' => $dist, 'prev' => $prev];
    }

    private function computeAllShortestPaths(): void
    {
        foreach ($this->cities as $a) {
            $result = $this->dijkstra($a);
            $this->distanceMatrix[$a] = $result['dist'];
            $this->pathMatrix[$a] = $result['prev'];
        }
    }

    public function getDistance(string $from, string $to): float
    {
        return $this->distanceMatrix[$from][$to] ?? INF;
    }

    public function getDistanceMatrix(): array { return $this->distanceMatrix; }
    public function getCities(): array { return $this->cities; }

    public function getEdges(): array
    {
        $edges = [];
        $seen = [];
        foreach ($this->graph as $u => $neighbors) {
            foreach ($neighbors as $n) {
                $v = $n['city'];
                $w = $n['weight'];
                $key = min($u, $v) . '|' . max($u, $v);
                if (!isset($seen[$key])) {
                    $seen[$key] = true;
                    $edges[] = ['from' => $u, 'to' => $v, 'weight' => $w];
                }
            }
        }
        return $edges;
    }

    /** Rekonstruksi path dari src ke dst */
    public function getPath(string $src, string $dst): array
    {
        if ($src === $dst) return [$src];
        $path = [];
        $current = $dst;
        $prev = $this->pathMatrix[$src] ?? [];
        while ($current !== null) {
            array_unshift($path, $current);
            if ($current === $src) break;
            $current = $prev[$current] ?? null;
        }
        return $path;
    }
}

File: services/graph_service.py
class GraphService:
    def __init__(self, edges):
        self.graph = {}
        self.cities = []
        self.distance_matrix = {}
        self.path_matrix = {}
        
        for edge in edges:
            u, v, w = edge
            if u not in self.graph:
                self.graph[u] = []
            if v not in self.graph:
                self.graph[v] = []
                
            self.graph[u].append({'city': v, 'weight': w})
            self.graph[v].append({'city': u, 'weight': w})
            
            if u not in self.cities:
                self.cities.append(u)
            if v not in self.cities:
                self.cities.append(v)
                
        self.compute_all_shortest_paths()

    def dijkstra(self, src):
        dist = {}
        prev = {}
        visited = {}
        for c in self.cities:
            dist[c] = float('inf')
            prev[c] = None
            visited[c] = False
        dist[src] = 0

        for _ in range(len(self.cities)):
            min_dist = float('inf')
            u = None
            for c in self.cities:
                if not visited[c] and dist[c] < min_dist:
                    min_dist = dist[c]
                    u = c
            
            if u is None:
                break
                
            visited[u] = True
            for neighbor in self.graph.get(u, []):
                alt = dist[u] + neighbor['weight']
                if alt < dist[neighbor['city']]:
                    dist[neighbor['city']] = alt
                    prev[neighbor['city']] = u
                    
        return {'dist': dist, 'prev': prev}

    def compute_all_shortest_paths(self):
        for a in self.cities:
            result = self.dijkstra(a)
            self.distance_matrix[a] = result['dist']
            self.path_matrix[a] = result['prev']

    def get_distance(self, from_city, to_city):
        if from_city in self.distance_matrix and to_city in self.distance_matrix[from_city]:
            return self.distance_matrix[from_city][to_city]
        return float('inf')

    def get_distance_matrix(self):
        return self.distance_matrix

    def get_cities(self):
        return self.cities

    def get_edges(self):
        edges = []
        seen = {}
        for u, neighbors in self.graph.items():
            for n in neighbors:
                v = n['city']
                w = n['weight']
                key = f"{min(u, v)}|{max(u, v)}"
                if key not in seen:
                    seen[key] = True
                    edges.append({'from': u, 'to': v, 'weight': w})
        return edges

    def get_path(self, src, dst):
        if src == dst:
            return [src]
        path = []
        current = dst
        prev_map = self.path_matrix.get(src, {})
        while current is not None:
            path.insert(0, current)
            if current == src:
                break
            current = prev_map.get(current, None)
        return path
